Python evaluator module fix

loomlike opened the pull request:
### Description
Python evaluation module' ranking metric functions have redundant and unnecessary sorting codes.
E.g. 
```
df_hit["rank"] = df_hit.groupby(col_user)[col_prediction].rank(
        method="first", ascending=False
)
```
doesn't need to use `rank()` since `df_hit` is already sorted by user and ratings as it is generated by groupby user (pandas groupby's `sort` argument is by default True) and nlargest ratings.

This change removes those redundant and unnecessary sorts and also refactor `get_top_k_items` to return DataFrame with 'rank' column to make its behavior the same as our pyspark evaluation module.

### Related Issues
<!--- If it fixes an open issue, please link to the issue here. -->


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.

gramhagen commented on reco_utils/evaluation/python_evaluation.py:
```diff
@@ -651,14 +648,16 @@ def get_top_k_items(
         .apply(lambda x: x.nlargest(k, col_rating))
         .reset_index(drop=True)
     )
+    top_k_items["rank"] = top_k_items.groupby(col_user).cumcount() + 1
```
you can avoid the repeated groupby too

```
groups = dataframe.groupby(col_user, as_index=False)
top_k_items = groups.apply(lambda x: x.nlargest(k, col_rating)).reset_index(drop=True)
top_k_items["rank"] = groups.cumcount() + 1
```

gramhagen commented on reco_utils/evaluation/python_evaluation.py:
```diff
@@ -651,14 +648,16 @@ def get_top_k_items(
 
     Returns:
-        pd.DataFrame: DataFrame of top k items for each user
+        pd.DataFrame: DataFrame of top k items for each user, sorted by `col_user` and `"rank"`
```
i would remove the double quotes from rank to match just the backticks like col_user

gramhagen replied:
also, in the returns section =)

loomlike replied:
good catch!

loomlike commented:
@gramhagen Few changes since the last review:
1. changed `"rank"` to `rank`
2. caching `groups` turns out does not work, since `nlargest` sorts the ratings while the cached group object still contains unsorted ratings. I changed it back to use groupby again, but added `sort=False` so that groupby can be performed efficiently (groupby-without-sorting still keeps the inter-group orders and we already sorted previously by 'nlargest')
3. found the above issue from spark's unit-tests which matches spark-evaluation-fn results to python's. Python evaluation tests couldn't catch the error because the test case users and items were already sorted. I made a simple tweak to the test case so that can catch such errors in the future.

gramhagen commented:
oh interesting, i didn't realize we use the python evaluation to validate test results for spark, we should remove that linkage, I'll add a separate feature request

gramhagen commented:
oh, i take it back, I guess that's an additional check just to ensure they match. i guess it helped in this case.

miguelgfierro commented:
@loomlike feel free to merge when you think it is convenient
